### PR TITLE
Part 3 of static analysis: clang-tidy

### LINKED ISSUE #1176: Static analysis meta-issue
Similar to #1166 I'm creating this issue to track warnings from compilers or static analysis.  Generally these aren't urgent to patch (I'll note if/when I think they need more immediate attention) but for future validation/PR workflows reducing the amount of these will allow us to use warnings that currently generate too many false-ish positives (true positives but probably not all that impactful) to be useful. 

To do this I'm currently building a custom version of the container environment with some additional tools (clang, clang-tidy so-far). If this turns out to be something we want to use in the future we could consider adding to the regular container. I'm also adding some additional support for adding additional warnings in the cmake module while working on https://github.com/LDMX-Software/cmake/issues/17. 

Whatever we decide for an approach for #1168 probably should guide if/how we would apply this.

Potential bugs/worth looking into

- [x] `-Wreturn-type` Control reaching the end of non-void functions. 
    - [x] GetGammaProcessManager (SimCore)
    - [x] getVisAttributes (SimCore)
    - [x] rotateGlobalToLocalBarPosition/getScintillatorOrientation (DetDescr)
- [x] `-Wmaybe-uninitialized` Using or recording uninitialized variables can be a real issue. 

Medium priority (fix when convenient)
- [x] Missing virtual for destructors. These are all over the place, mostly in processors, and probably not an issue most of the time. However, if you have a destructor that actually needs to do something then this could be a problem. This seems straight-forward to patch whenever you are touching a corresponding part of the code 
- [x] Shadowed virtual functions (e.g. `virtual void onFileOpen() {}` instead of `virtual void onFileOpen(EventFile &eventFile) {}`
- [x] `-Wunused` Unused variables. Often just left over from development but can also be a mistake. 
- [x] `-Wreorder`: C++ member variables are always initialized in the order they are declared, even if the constructor initializer list has a different order. Usually harmless but if there are any dependencies on the order of parameters then it can be a big problem 

Low priority (might not be worth fixing)
- [x] Signed/unsigned int conversions. These are also all over the place but I don't think these need to be patched in the same way. It comes up a lot when dealing with DetectorIDs (where most things are unsigned internally) 
- [x] Double promotion. This I suspect is even less important than the signed/unsigned int conversions, we can probably just turn this off (note: this is not narrowing from double to float, which could be an issue) 
- [ ]  

Related issues/PRs
- [x] https://github.com/LDMX-Software/ldmx-sw/issues/1208 
- [x] https://github.com/LDMX-Software/Framework/issues/70 
- [x] https://github.com/LDMX-Software/ldmx-sw/issues/1199, https://github.com/LDMX-Software/ldmx-sw/pull/1207
- [x] https://github.com/LDMX-Software/ldmx-sw/issues/1172 

PRs with warning cleanup as part of the PR 
- [x] https://github.com/LDMX-Software/Hcal/pull/70 
- [x] https://github.com/LDMX-Software/TrigScint/pull/54 
- [x] https://github.com/LDMX-Software/SimCore/pull/95

### PR DESCRIPTION
**Pull request by tvami**
I am updating _ldmx-sw_, here are the details.

### What are the issues that this addresses?
Resolves https://github.com/LDMX-Software/ldmx-sw/issues/1176

## Check List
- [x] I successfully compiled _ldmx-sw_ with my developments
- [x] I ran my developments and the following shows that they are successful.

I killed all the warning. 
```
rm -rf build
just configure -DADDITIONAL_WARNINGS=ON  -DENABLE_CLANG_TIDY=ON  -DENABLE_LTO=ON  -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_C_COMPILER=clang
just build
// And switch back to a build with GCC
rm -rf build 
just configure -DADDITIONAL_WARNINGS=ON  -DENABLE_CLANG_TIDY=ON  -DENABLE_LTO=ON 
just build
```
is now nice and clean.

### CODE REVIEW

**Review by danyi211 — APPROVED** [review]

**Comment by tvami** [comment]
Tests ran fine https://github.com/LDMX-Software/ldmx-sw/actions/runs/10894469361
(except for the ecal pn, but that's broken bc of the tracking change, I'll fix it with https://github.com/LDMX-Software/ldmx-sw/issues/1459 tomorrow)

File: CMakeLists.txt [hunk 1]
```diff
@@ -36,6 +36,9 @@ include_directories(SYSTEM /usr/local/include/root)
 include_directories(SYSTEM Tracking/acts/Core/include)
 include_directories(SYSTEM Trigger/HLS_arbitrary_Precision_Types/include)
 
+# FunctionalCoreTest.cxx doesnt comply with clang-tidy but that's ok
+set_source_files_properties(Framework/test/FunctionalCoreTest.cxx PROPERTIES COMPILE_OPTIONS "-Wno-null-dereference")
+
```
Comment (tomeichlersmith) [inline CMakeLists.txt:41]: Is it possible to move this into `Framework/CMakeLists.txt`?

File: Framework/src/Framework/Event.cxx [hunk 2]
```diff
@@ -121,8 +121,15 @@ void Event::setInputTree(TTree* tree) {
 
   // find the names of all the existing branches
   TObjArray* branches = inputTree_->GetListOfBranches();
+  if (!branches) {
+    EXCEPTION_RAISE("RunHeaderError", "Branch doesnt exists");
```
Comment (tomeichlersmith) [inline Framework/src/Framework/Event.cxx]:
```suggestion
    EXCEPTION_RAISE("BadInputFile", "Input tree doesn't have a list of branches but still made it to this point. This is bad and has never been seen before!");
```

File: justfile [hunk 3]
```diff
@@ -59,7 +59,13 @@ install-denv:
 
 # configure how ldmx-sw will be built
 configure *CONFIG:
-    denv cmake -B build -S . {{ CONFIG }}
+    denv cmake -B build -S .  -DADDITIONAL_WARNINGS=ON  -DENABLE_CLANG_TIDY=ON  {{ CONFIG }}
+
+configure-force-error *CONFIG:
+    just  configure -DWARNINGS_AS_ERRORS=ON {{ CONFIG }}
```
Comment (tomeichlersmith) [inline justfile]:
```suggestion
configure-force-error *CONFIG: (configure -DWARNINGS_AS_ERRORS=ON CONFIG)
```

File: justfile [hunk 4]
```diff
@@ -59,7 +59,13 @@ install-denv:
...
+
+configure-clang-lto:
+    denv cmake -B build -S . -DADDITIONAL_WARNINGS=ON  -DENABLE_CLANG_TIDY=ON  -DENABLE_LTO=ON  -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_C_COMPILER=clang
```
Comment (tomeichlersmith) [inline justfile]: Is falling back to `denv` necessary here?

```suggestion
configure-clang-lto: (configure -DENABLE_LTO=ON -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_C_COMPILER=clang) 
```

File: justfile [hunk 5]
```diff
@@ -59,7 +59,13 @@ install-denv:
 
 # configure how ldmx-sw will be built
 configure *CONFIG:
-    denv cmake -B build -S . {{ CONFIG }}
+    denv cmake -B build -S .  -DADDITIONAL_WARNINGS=ON  -DENABLE_CLANG_TIDY=ON  {{ CONFIG }}
```
Comment (tomeichlersmith) [inline justfile:63]: If we really want these CMake options to be ON by default, maybe we should put them into the CMakeLists.txt like that?

e.g. I think the following would still allow us to disable on the command line but would change the default to `ON` if not specified.

```cmake
if (NOT ADDITIONAL_WARNINGS)
  set(ADDITIONAL_WARNINGS ON)
endif()
```

**Review by tomeichlersmith — COMMENTED** [review]
C++ changes look good, my comments are mainly focused on passing arguments to recipes within the justfile simply to avoid the recursive call to the program itself and maybe moving default build settings into the CMakeLists.txt.

File: CMakeLists.txt (hunk 1 above)
Comment (tvami) [reply CMakeLists.txt:41]: I tried this, it does not work! :(
EDIT: oh maybe it's that I copied the whole thing and not think about the relative path

Comment (tvami) [reply CMakeLists.txt:41]: Nope, it doesnt work with the below either
```
set_source_files_properties(test/FunctionalCoreTest.cxx PROPERTIES COMPILE_OPTIONS "-Wno-null-dereference")
```
I'm going back to the original

File: Framework/src/Framework/Event.cxx (hunk 2 above)
Comment (tvami) [reply Framework/src/Framework/Event.cxx]: done in the next commit

File: justfile (hunk 3 above)
Comment (tvami) [reply justfile]: Implemented it as `configure-force-error *CONFIG: (configure "-DWARNINGS_AS_ERRORS=ON" CONFIG)`

Comment (tomeichlersmith) [reply justfile]: We found out that `just` treats `-` as tokens within its own syntax and so we need to wrap the `-DWARNINGS_AS_ERRORS=ON` with `"` to inform `just` that it should be treated as an argument to the recipe and not apart of its own syntax.

File: justfile (hunk 4 above)
Comment (tvami) [reply justfile]: Done as 
```
configure-clang-lto: (configure "-DENABLE_LTO=ON -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_C_COMPILER=clang")
```

File: justfile (hunk 5 above)
Comment (tvami) [reply justfile:63]: So it's yes and no. It does seem to work, but gives this other warning
```
CMake Warning (dev) at cmake/CompilerWarnings.cmake:13 (option):
  Policy CMP0077 is not set: option() honors normal variables.  Run "cmake
  --help-policy CMP0077" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  For compatibility with older versions of CMake, option is clearing the
  normal variable 'ADDITIONAL_WARNINGS'.
Call Stack (most recent call first):
  cmake/BuildMacros.cmake:143 (enable_compiler_warnings)
  Recon/CMakeLists.txt:37 (setup_library)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cmake/StaticAnalysis.cmake:5 (option):
  Policy CMP0077 is not set: option() honors normal variables.  Run "cmake
  --help-policy CMP0077" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  For compatibility with older versions of CMake, option is clearing the
  normal variable 'ENABLE_CLANG_TIDY'.
Call Stack (most recent call first):
  cmake/BuildMacros.cmake:145 (enable_clang_tidy)
  Recon/CMakeLists.txt:37 (setup_library)
This warning is for project developers.  Use -Wno-dev to suppress it.
```
which makes it pretty ugly... so I'll just keep this as it was. Honestly I think it's more transparent too to have it explicitly like this in the justfile

**Review by tomeichlersmith — APPROVED** [review]

**Pull request merged**